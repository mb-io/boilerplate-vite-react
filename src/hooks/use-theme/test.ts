import { describe, it, expect, beforeEach } from 'vitest'
import { useTheme, ThemeProps } from './use-theme'
import { act, renderHook } from '@testing-library/react'

describe('useTheme', () => {
  beforeEach(() => {
    localStorage.clear()
    document.documentElement.className = ''
  })

  it('should initialize with default dark theme', () => {
    const { result } = renderHook(() => useTheme())
    expect(result.current.theme).toBe(ThemeProps.dark)
    expect(document.documentElement.classList.contains(ThemeProps.dark)).toBe(true)
    expect(result).toMatchSnapshot()
  })

  it('should toggle theme', () => {
    const { result } = renderHook(() => useTheme())
    act(() => {
      result.current.toggleTheme()
    })
    expect(result.current.theme).toBe(ThemeProps.light)
    expect(document.documentElement.classList.contains(ThemeProps.light)).toBe(true)
    expect(result).toMatchSnapshot()
  })

  it('should set light theme', () => {
    const { result } = renderHook(() => useTheme())
    act(() => {
      result.current.setLightTheme()
    })
    expect(result.current.theme).toBe(ThemeProps.light)
    expect(document.documentElement.classList.contains(ThemeProps.light)).toBe(true)
    expect(result).toMatchSnapshot()
  })

  it('should set dark theme', () => {
    const { result } = renderHook(() => useTheme())
    act(() => {
      result.current.setDarkTheme()
    })
    expect(result.current.theme).toBe(ThemeProps.dark)
    expect(document.documentElement.classList.contains(ThemeProps.dark)).toBe(true)
    expect(result).toMatchSnapshot()
  })
})
