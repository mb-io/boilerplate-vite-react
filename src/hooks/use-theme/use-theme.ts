import { useEffect, useMemo, useState } from 'react'

import { useThemeStore } from '@/stores/theme-store/theme-store'

export const ThemeProps = {
  key: 'theme',
  light: 'light',
  dark: 'dark'
} as const

export type Theme = typeof ThemeProps.light | typeof ThemeProps.dark

export const useTheme = (defaultTheme?: Theme) => {
  const updateTheme = useThemeStore((state) => state.updateTheme)
  const [isMounted, setIsMounted] = useState(false)
  const [theme, setTheme] = useState<Theme>(() => {
    const storedTheme = localStorage.getItem(ThemeProps.key) as Theme | null

    return storedTheme ?? defaultTheme ?? ThemeProps.dark
  })

  const isDark = useMemo(() => {
    return theme === ThemeProps.dark
  }, [theme])

  const isLight = useMemo(() => {
    return theme === ThemeProps.light
  }, [theme])

  const _setTheme = (nextTheme: Theme) => {
    localStorage.setItem(ThemeProps.key, nextTheme)
    document.documentElement.classList.remove(ThemeProps.light, ThemeProps.dark)
    document.documentElement.classList.add(nextTheme)
    setTheme(nextTheme)
  }

  const setLightTheme = () => _setTheme(ThemeProps.light)

  const setDarkTheme = () => _setTheme(ThemeProps.dark)

  const toggleTheme = () => {
    if (theme === ThemeProps.dark) {
      setLightTheme()
      updateTheme(ThemeProps.light)
    } else {
      setDarkTheme()
      updateTheme(ThemeProps.dark)
    }
  }

  useEffect(() => {
    _setTheme(theme)
    if (!isMounted) updateTheme(theme)
    setIsMounted(true)
  }, [theme, isMounted, updateTheme])

  return { theme, isDark, isLight, setLightTheme, setDarkTheme, toggleTheme }
}

export default useTheme
