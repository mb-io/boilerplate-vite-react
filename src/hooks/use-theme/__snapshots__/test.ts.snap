// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTheme > should initialize with default dark theme 1`] = `
{
  "current": {
    "isDark": true,
    "isLight": false,
    "setDarkTheme": [Function],
    "setLightTheme": [Function],
    "theme": "dark",
    "toggleTheme": [Function],
  },
}
`;

exports[`useTheme > should set dark theme 1`] = `
{
  "current": {
    "isDark": true,
    "isLight": false,
    "setDarkTheme": [Function],
    "setLightTheme": [Function],
    "theme": "dark",
    "toggleTheme": [Function],
  },
}
`;

exports[`useTheme > should set light theme 1`] = `
{
  "current": {
    "isDark": false,
    "isLight": true,
    "setDarkTheme": [Function],
    "setLightTheme": [Function],
    "theme": "light",
    "toggleTheme": [Function],
  },
}
`;

exports[`useTheme > should toggle theme 1`] = `
{
  "current": {
    "isDark": false,
    "isLight": true,
    "setDarkTheme": [Function],
    "setLightTheme": [Function],
    "theme": "light",
    "toggleTheme": [Function],
  },
}
`;
